### PR TITLE
Ensure JDWP threads are not halted when hooks run for CRIU

### PR DESCRIPTION
**Pull request by mikezhang1234567890**
A breakpoint set during a pre-checkpoint or post-restore hook will cause Java to hang because the JDWP threads are suspended at that time. This change will delay the suspending of JDWP threads for pre-checkpoint hooks and resume them early for post-restore hooks if VM is run with the JDWP agentlib loaded.

JDK17 sanity.functional testing running internally [here.](https://hyc-runtimes-jenkins.swg-devops.com/job/Pipeline-Build-Test-Personal/15437/)

fyi @tajila @JasonFengJ9

### CODE REVIEW

File: runtime/criusupport/criusupport.cpp [hunk 1]
```diff
@@ -748,7 +783,7 @@ Java_org_eclipse_openj9_criu_CRIUSupport_checkpointJVMImpl(JNIEnv *env,
 
 		vm->extendedRuntimeFlags2 &= ~J9_EXTENDED_RUNTIME2_CRIU_SINGLE_THREAD_MODE;
 
-		toggleSuspendOnJavaThreads(currentThread, FALSE);
+		toggleSuspendOnJavaThreads(currentThread, FALSE, isJdwpEnabled);
```
Comment (JasonFengJ9) [inline runtime/criusupport/criusupport.cpp:786]: Does this clear the halt flag of non-JDWP threads when JDWP is enabled?

Comment (mikezhang1234567890) [reply runtime/criusupport/criusupport.cpp:786]: Right, the third parameter should always be `FALSE` for this line since it should take the default path of clearing every thread.

File: runtime/criusupport/criusupport.cpp [hunk 2]
```diff
@@ -443,6 +464,7 @@ Java_org_eclipse_openj9_criu_CRIUSupport_checkpointJVMImpl(JNIEnv *env,
 		char *syslogOptions = NULL;
 		I_32 syslogBufferSize = 0;
 		UDATA oldVMState = VM_VMHelpers::setVMState(currentThread, J9VMSTATE_CRIU_SUPPORT_CHECKPOINT_PHASE_START);
+		BOOLEAN isJdwpEnabled = FIND_ARG_IN_VMARGS(STARTSWITH_MATCH, "-agentlib:jdwp", NULL) >= 0;
```
Comment (tajila) [inline runtime/criusupport/criusupport.cpp:467]: add a state variable to `J9CRIUCheckpointState` and do the check on `setupJNIFieldIDsAndCRIUAPI` instead

File: runtime/criusupport/criusupport.cpp [hunk 3]
```diff
@@ -245,11 +247,26 @@ getNativeString(J9VMThread *currentThread, j9object_t javaString, char **nativeS
...
+	char *threadName = getOMRVMThreadName(currentThread->omrVMThread);
+	return 0 == strncmp("JDWP", threadName, 4);
+}
+
 /**
  * Caller must first acquire exclusive safepoint or exclusive VMAccess
  */
 static void
-toggleSuspendOnJavaThreads(J9VMThread *currentThread, BOOLEAN suspend)
+toggleSuspendOnJavaThreads(J9VMThread *currentThread, BOOLEAN suspend, BOOLEAN jdwpSpecialAction)
```
Comment (tajila) [inline runtime/criusupport/criusupport.cpp:269]: perhaps  `toggleDebuggingThreads` is more descriptive than `jdwpSpecialAction`

File: runtime/criusupport/criusupport.cpp [hunk 4]
```diff
@@ -262,9 +279,13 @@ toggleSuspendOnJavaThreads(J9VMThread *currentThread, BOOLEAN suspend)
 		&& (currentThread != walkThread)
 		) {
 			if (suspend) {
-				vmFuncs->setHaltFlag(walkThread, J9_PUBLIC_FLAGS_HALT_THREAD_FOR_CHECKPOINT);
+				if (!jdwpSpecialAction || (jdwpSpecialAction && !isJdwpDebugThread(walkThread))) {
```
Comment (tajila) [inline runtime/criusupport/criusupport.cpp:282]: you can define `bool skipThread = ...` above to avoid code duplication

**Comment by mikezhang1234567890** [comment]
Putting this in draft since I ran into an issue with threads ending up in inconsistent state after the restore, which causes a hang.

**Comment by mikezhang1234567890** [comment]
The hang was caused by the finalizer trying to create a worker thread during the pre-checkpoint hooks, however, the calls to `attachSystemDaemonThread` and `initializeAttachedThread` in the creation of the worker thread only happened post restore, causing a hang.

Delaying actually running the finalizer until after the restore is done and the VM exits single threaded mode seems to fix this issue.

@amicic @dmitripivkine do you have any problems with this approach of delaying the finalizer?

Also can you take another look please? @tajila

File: runtime/criusupport/criusupport.cpp (hunk 2 above)
Comment (mikezhang1234567890) [reply runtime/criusupport/criusupport.cpp]: Added in https://github.com/eclipse-openj9/openj9/pull/16653/commits/15a2fc49467422d926e8a58659658e35876c4e14

File: runtime/criusupport/criusupport.cpp (hunk 3 above)
Comment (mikezhang1234567890) [reply runtime/criusupport/criusupport.cpp]: Changed in https://github.com/eclipse-openj9/openj9/pull/16653/commits/15a2fc49467422d926e8a58659658e35876c4e14

File: runtime/criusupport/criusupport.cpp (hunk 4 above)
Comment (mikezhang1234567890) [reply runtime/criusupport/criusupport.cpp]: Added a boolean, but changed the logic a bit so that the function will only set or clear bits for JDWP threads or non JDWP threads.

**Comment by dmitripivkine** [comment]
Well, we can delay the finalizer if there is necessary. It depends what Finalization expectation is. @tajila Do we need Finalization to perform in `J9_EXTENDED_RUNTIME2_CRIU_SINGLE_THREAD_MODE`?

**Comment by tajila** [comment]
> Well, we can delay the finalizer if there is necessary. It depends what Finalization expectation is. @tajila Do we need Finalization to perform in J9_EXTENDED_RUNTIME2_CRIU_SINGLE_THREAD_MODE?

The mutator will be running in single thread mode. We dont need finalization to happen during this mode, but we need it to continue after single thread mode.

**Comment by dmitripivkine** [comment]
> > Well, we can delay the finalizer if there is necessary. It depends what Finalization expectation is. @tajila Do we need Finalization to perform in J9_EXTENDED_RUNTIME2_CRIU_SINGLE_THREAD_MODE?
> 
> The mutator will be running in single thread mode. We dont need finalization to happen during this mode, but we need it to continue after single thread mode.

It will. Returning negative from `FinalizeMainRunFinalization()` causes free spinning in the `do` loop in `FinalizeMainThread()`. Free spinning can be a problem potentially. One of known problems is ZOS related (and not applicable for CRIU obviously): free spinning cause 100% CPU utilization blocking execution by other threads. I am not sure what negative impact can be introduced on Linux platforms. So, we have a choice - keep it free spinning "as is" now in this change (and anticipate potential problems) or introduce more advanced handling (for example trying to wait on Monitor)

**Comment by dmitripivkine** [comment]
There is some waiting mechanism exists already, trying to figure out how it can be reused:
https://github.com/eclipse-openj9/openj9/blob/9bf20a075d01f0f2377a15c0adad686aa648259c/runtime/gc_base/FinalizerSupport.cpp#L222:#L233

**Comment by mikezhang1234567890** [comment]
Returning negative in `FinalizeMainRunFinalization()` sets `currentWaitTime` and `noCycleWait` to 0 so every iteration should hit the wait at https://github.com/eclipse-openj9/openj9/blob/9bf20a075d01f0f2377a15c0adad686aa648259c/runtime/gc_base/FinalizerSupport.cpp#L229. Would we still need another waiting loop based on the `J9_EXTENDED_RUNTIME2_CRIU_SINGLE_THREAD_MODE`?

**Comment by dmitripivkine** [comment]
Yes, you are may be correct but is depends on flag set to `vm->finalizeMainFlags` to wakeup Main Finalization thread. Is it `J9_FINALIZE_FLAGS_MAIN_WAKE_UP`? Would you please provide value?

**Comment by amicic** [comment]
I don't think there is correctness problem with deferring creation of the finalization worker thread and doing any finalization that was created by last GC in checkpoint, although for perf (short restore times) it would be better to complete all that on checkpoint side. But that approach could be non-trivial (finalization is its nature async and we generally don't know when it's completed) and I would not consider it now.

As for how we defer the creation of the worker thread, seems like it would be better to not allow exiting the wait loop that @dmitripivkine pointed to, under the same CRIU single threaded condition that you used. Not exiting from the loop would keep all the finalization flags and misc parameters intact, till the CRIU condition is cleared up.

The loop already has a timeout, so we will eventually wake up during restore, and in fact in real life probably sooner (1/2 sec) then waiting for a new worker thread wake up flag is set (arbitrary time till next GC that has finalization work) in case we take the current approach (let wake up flag be cleared and then check CRIU condition).

From what I can see the single thread mode is set even on checkpoint side, even before we call j9gc_prepare_for_checkpoint, which is actually good, so it would help prevent a race between creation of the finalization worker thread vs checkpoint and restore activity.

**Comment by mikezhang1234567890** [comment]
Changing it to use `omrthread_monitor_wait_timed` loop in https://github.com/eclipse-openj9/openj9/commit/9cb649e2b24e50cd71ca3d9e039202ae3e24f91a caused the hang to reappear.

Also, I found out using the `threads` command in `jdb` also causes a hang, during hooks and after the CRIU restore is complete. This occurs even when using `-Xgcpolicy:nogc`.

What's common between the finalizer thread and the jdb get threads command is they both call `getOMRVMThreadName` which seems to get stuck waiting at `omrthread_monitor_enter(vmThread->threadNameMutex);`

**Comment by mikezhang1234567890** [comment]
The problem turned out to be a missing `releaseOMRVMThreadName` call after `getOMRVMThreadName` in `isJdwpDebugThread`, so we never exited the monitor. 

Now no longer makes changes to GC code in https://github.com/eclipse-openj9/openj9/commit/253019c353737a934c46d0f4e63ec501e5a4eb5b

@tajila can you take another look please?

File: runtime/criusupport/CMakeLists.txt [hunk 5]
```diff
@@ -45,6 +45,7 @@ target_link_libraries(j9criu
 		dl
 		j9utilcore
 		j9pool
+		omrutil
```
Comment (JasonFengJ9) [inline runtime/criusupport/CMakeLists.txt:48]: Is this change required?

File: runtime/criusupport/criusupport.cpp [hunk 6]
```diff
@@ -143,6 +145,8 @@ setupJNIFieldIDsAndCRIUAPI(JNIEnv *env, jclass *currentExceptionClass, IDATA *sy
 		goto done;
 	}
 
+	vm->checkpointState.isJdwpEnabled = FIND_ARG_IN_VMARGS(STARTSWITH_MATCH, "-agentlib:jdwp", NULL) >= 0;
```
Comment (JasonFengJ9) [inline runtime/criusupport/criusupport.cpp:148]: Can `MAPOPT_AGENTLIB_JDWP_EQUALS "-agentlib:jdwp="` be used here?

File: runtime/criusupport/criusupport.cpp [hunk 7]
```diff
@@ -244,11 +248,31 @@ getNativeString(J9VMThread *currentThread, j9object_t javaString, char **nativeS
...
+/**
+ * Checks if a thread is used by JDWP for debugging. Use string compare to check that the thread
+ * name starts with "JDWP" since all JDWP threads have that as a prefix.
+ *
+ * @param[in] currentThread the thread being checked
+ *
+ * @return true if the thread is used by JDWP, false if not
+ */
+static bool
+isJdwpDebugThread(J9VMThread *currentThread)
```
Comment (JasonFengJ9) [inline runtime/criusupport/criusupport.cpp:260]: This can be expanded to take an additional parameter `toggleDebugThreads` such that it contains the logic of `((!toggleDebugThreads && !isJdwpDebugThread(walkThread)) || (toggleDebugThreads && isJdwpDebugThread(walkThread)))`, and simplify the `toggleThread` check within following `toggleSuspendOnJavaThreads`.

File: runtime/gc_base/FinalizerSupport.cpp [hunk 8]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 1991, 2022 IBM Corp. and others
+ * Copyright (c) 1991, 2023 IBM Corp. and others
```
Comment (JasonFengJ9) [inline runtime/gc_base/FinalizerSupport.cpp:2]: This change can be removed.

File: runtime/criusupport/CMakeLists.txt (hunk 5 above)
Comment (mikezhang1234567890) [reply runtime/criusupport/CMakeLists.txt:48]: It's needed for `getOMRVMThreadName` and `releaseOMRVMThreadName`

File: runtime/criusupport/criusupport.cpp (hunk 6 above)
Comment (tajila) [reply runtime/criusupport/criusupport.cpp:148]: I dont think `-agentlib:jdwp=` is enough, you can enable debugging support with `-Xrunjdwp` as well

Comment (mikezhang1234567890) [reply runtime/criusupport/criusupport.cpp:148]: Ya looks like `-Xrunjdwp` also needs to be checked. I thought it was mapped to the first option.

Comment (JasonFengJ9) [reply runtime/criusupport/criusupport.cpp:148]: Something was broken, or following comment is obsolated
https://github.com/eclipse-openj9/openj9/blob/59bdae92b1d3026620b2601275d0b3da14690334/runtime/vm/jvminit.c#L1418

File: runtime/criusupport/criusupport.cpp [hunk 9]
```diff
@@ -592,6 +622,13 @@ Java_org_eclipse_openj9_criu_CRIUSupport_checkpointJVMImpl(JNIEnv *env,
 			/* throw the pending exception */
 			goto wakeJavaThreads;
 		}
+
+		if (vm->checkpointState.isJdwpEnabled) {
+			acquireSafeOrExcusiveVMAccess(currentThread, vmFuncs, safePoint);
```
Comment (tajila) [inline runtime/criusupport/criusupport.cpp:627]: this is done too early, shutting down GC threads could trigger a JVMTI event, if all JVMTI threads are blocked then this will lead to a deadlock. This happens fairly consistenly on my machine.

The JVMTI threads need to be shutdown after `runInternalJVMCheckpointHooks` is called

File: runtime/criusupport/CMakeLists.txt (hunk 5 above)
Comment (mikezhang1234567890) [reply runtime/criusupport/CMakeLists.txt:48]: Looks like it's not needed anymore now that j9utilcore is added as a target.

File: runtime/criusupport/criusupport.cpp [hunk 10]
```diff
@@ -592,6 +622,13 @@ Java_org_eclipse_openj9_criu_CRIUSupport_checkpointJVMImpl(JNIEnv *env,
...
+			toggleSuspendOnJavaThreads(currentThread, TRUE, TRUE);
+			releaseSafeOrExcusiveVMAccess(currentThread, vmFuncs, safePoint);
+		}
+
```
Comment (tajila) [inline runtime/criusupport/criusupport.cpp:631]: youill need to add a vm release/acquire in `j9gc_prepare_for_checkpoint` as well

```
	releaseVMAccess(vmThread);
	extensions->configuration->adjustGCThreadCountForCheckpoint(env);
	acquireVMAccess(vmThread);
```

File: runtime/criusupport/criusupport.cpp (hunk 6 above)
Comment (mikezhang1234567890) [reply runtime/criusupport/criusupport.cpp]: Added a check for `Xrunjdwp` as well for setting `isJdwpEnabled`

File: runtime/criusupport/criusupport.cpp (hunk 9 above)
Comment (mikezhang1234567890) [reply runtime/criusupport/criusupport.cpp]: Moved the disabling of JDWP threads after the internal hooks before checkpoint, and re-enabling of JDWP threads to before the GC reinitializes threads after restore.

**Comment by mikezhang1234567890** [comment]
Updated PR to address review comments, can you please take another look @JasonFengJ9 @tajila

**Comment by tajila** [comment]
@dmitripivkine please review GC changes

**Comment by dmitripivkine** [comment]
> @dmitripivkine please review GC changes

:face_with_monocle:

File: runtime/gc_base/modronapi.cpp [hunk 11]
```diff
@@ -1116,7 +1116,9 @@ j9gc_prepare_for_checkpoint(J9VMThread *vmThread)
 		verboseGCManager->prepareForCheckpoint(env);
 	}
 
+	releaseVMAccess(vmThread);
```
Comment (dmitripivkine) [inline runtime/gc_base/modronapi.cpp:1119]: Would you please add a comment why VM Access should be released? Something like: "threads being terminated may trigger a hook that may acquire VM access via JVMTI callback"

**Review by dmitripivkine — APPROVED** [review]
ok, please add comment

File: runtime/gc_base/modronapi.cpp (hunk 11 above)
Comment (mikezhang1234567890) [reply runtime/gc_base/modronapi.cpp:1120]: Added comment with this explanation.

Comment (amicic) [reply runtime/gc_base/modronapi.cpp:1120]: may acquire exclusive VM access

Comment (mikezhang1234567890) [reply runtime/gc_base/modronapi.cpp:1120]: Updated wording in https://github.com/eclipse-openj9/openj9/commit/6ab4114e3a49f0db2f43f56f6945de2bd6ad8fd3

File: runtime/criusupport/criusupport.cpp [hunk 12]
```diff
@@ -143,6 +145,9 @@ setupJNIFieldIDsAndCRIUAPI(JNIEnv *env, jclass *currentExceptionClass, IDATA *sy
 		goto done;
 	}
 
+	vm->checkpointState.isJdwpEnabled = FIND_ARG_IN_VMARGS(STARTSWITH_MATCH, MAPOPT_AGENTLIB_JDWP_EQUALS, NULL) >= 0
+		|| FIND_ARG_IN_VMARGS(STARTSWITH_MATCH, MAPOPT_XRUNJDWP, NULL) >= 0;
```
Comment (JasonFengJ9) [inline runtime/criusupport/criusupport.cpp:149]:
```suggestion
vm->checkpointState.isJdwpEnabled = (FIND_ARG_IN_VMARGS(STARTSWITH_MATCH, MAPOPT_AGENTLIB_JDWP_EQUALS, NULL) >= 0)
		|| (FIND_ARG_IN_VMARGS(STARTSWITH_MATCH, MAPOPT_XRUNJDWP, NULL) >= 0);
```

File: runtime/criusupport/criusupport.cpp [hunk 13]
```diff
@@ -245,10 +250,43 @@ getNativeString(J9VMThread *currentThread, j9object_t javaString, char **nativeS
...
+		char *threadName = getOMRVMThreadName(currentThread->omrVMThread);
+		releaseOMRVMThreadName(currentThread->omrVMThread);
+		if (toggleDebugThreads) {
+			/* all threads started by JDWP begin with "JDWP" in their name */
+			return 0 == strncmp("JDWP", threadName, 4);
+		} else {
+			return 0 != strncmp("JDWP", threadName, 4);
+		}
+	}
+	return true;
```
Comment (JasonFengJ9) [inline runtime/criusupport/criusupport.cpp:277]: I would suggest the following to avoid duplicated `strncmp` calls and have a single return point.
```
	J9JavaVM *vm = currentThread->javaVM;
	bool result = true;
	if (vm->checkpointState.isJdwpEnabled) {
		char *threadName = getOMRVMThreadName(currentThread->omrVMThread);
		releaseOMRVMThreadName(currentThread->omrVMThread);
		/* all threads started by JDWP begin with "JDWP" in their name */
		bool isJDWPThread = 0 == strncmp("JDWP", threadName, 4);
		if (toggleDebugThreads) {
			result = isJDWPThread;
		} else {
			result = !isJDWPThread;
		}
	}
	return result;
```

File: runtime/criusupport/criusupport.cpp (hunk 12 above)
Comment (mikezhang1234567890) [reply runtime/criusupport/criusupport.cpp]: Changed in https://github.com/eclipse-openj9/openj9/commit/dfaac2f52d0ef19bf0b989cc66070b47bdcef452

**Review by JasonFengJ9 — APPROVED** [review]
LGTM

**Review by tajila — APPROVED** [review]

**Comment by tajila** [comment]
jenkins compile win jdk8

**Comment by tajila** [comment]
jenkins test sanity xlinux jdk17

**Comment by tajila** [comment]
@mikezhang1234567890 please open a PR for the 0.38 branch
